File: async/javastore/app/models/StockQuery.java
package models;

import org.codehaus.jackson.JsonNode;
import play.libs.F;
import play.libs.WS;

import static play.libs.F.Promise;

public class StockQuery {

    public Promise<JsonNode> searchStock(String query) {
        Promise<WS.Response> responsePromise = WS.url("http://localhost:9001/search").setQueryParameter("query", query).get();
        return responsePromise.map(new F.Function<WS.Response, JsonNode>() {
            @Override
            public JsonNode apply(WS.Response response) throws Throwable {
                if (response.getStatus() == 200) {
                    JsonNode jsonNode = response.asJson();
                    return jsonNode.findPath("results");
                } else {
                    throw new Exception("Error calling search service.\nResponse status " + response.getStatus() + "\n");
                }
            }
        });
    }
}

File: async/scalastore/app/models/StockQuery.scala
package models

import play.api.libs.ws.WS
import play.api.libs.json.JsObject
import scala.util.{Failure, Try}
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

trait StockQuery {

  def searchStock(query: String): Future[Try[Seq[JsObject]]] = {
    WS.url("http://localhost:9001/search").withQueryString("query" -> query).get().map { response =>
      response.status match {
        case 200 =>
          Try((response.json \ "results").as[Seq[JsObject]])

        case _ => Failure(new Exception(s"Error calling search service.\nResponse status ${response.status}\n"))
      }
    }
  }

}
